bpo-31540 Add context management for concurrent.futures.ProcessPoolExecutor

tomMoral opened the pull request:
The `ProcessPoolExecutor` processes start method can only be change by changing the global default context with `set_start_method` at the beginning of a script. We propose to allow passing a context argument in the constructor to allow more flexible control of the executor. Doing so, we also add some tests for all the available context, to make sure the executor is working correctly.

In addition, we made the following changes, which can be put in another PR if necessary:

* Rename the global variable `_shutdown` to `_global_shutdown` to make its function in the code clearer.
* Liberate the ressource earlier in the `_worker_process`. Indeed, with the actual behavior, the ressources are not freed before the `worker` receives a new task or shutdown.

This work was done as part of the [`loky`][1] project in collaboration with
@ogrisel. See #1013 for the details.

[1]: https://github.com/tomMoral/loky

<!-- issue-number: bpo-31540 -->
https://bugs.python.org/issue31540
<!-- /issue-number -->

tomMoral committed:
Add context management for ProcessPoolExecutor+CLN

tomMoral/loky#48
* Add context argument to allow non forking ProcessPoolExecutor
* Do some cleaning (pep8+nonused code+naming)
* Liberate the ressource earlier in the `_worker_process`

tomMoral committed:
FIX skip tests that re-run the entire test suite

tomMoral committed:
NEW add whatsnew entry

pitrou commented on Lib/concurrent/futures/process.py:
```diff
@@ -180,6 +177,11 @@ def _process_worker(call_queue, result_queue):
             result_queue.put(_ResultItem(call_item.work_id,
                                          result=r))
 
+        # Liberate the resource as soon as possible, to avoid holding onto
```
Would it be easy to add a test for this?

tomMoral replied:
I added a test for this behavior.

pitrou commented on Lib/concurrent/futures/process.py:
```diff
@@ -377,13 +380,15 @@ class BrokenProcessPool(RuntimeError):
             max_workers: The maximum number of processes that can be used to
                 execute the given calls. If None or not given then as many
                 worker processes will be created as the machine has processors.
+            context: A multiprocessing context to launch the workers. This
```
As a nit, I think calling this parameter `mp_context` would be a bit more explicit.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -123,8 +151,9 @@ def test_interpreter_shutdown(self):
-            t = {executor_type}(5)
-            t.submit(sleep_and_print, 1.0, "apple")
+            if __name__ == "__main__":
+                t = {executor_type}(5)
```
Perhaps we want to pass the right context argument here?

pitrou commented on .travis.yml:
```diff
@@ -56,7 +56,7 @@ matrix:
       script:
         # Skip tests that re-run the entire test suite.
-        - ./venv/bin/python -m coverage run --pylib -m test --fail-env-changed -uall,-cpu -x test_multiprocessing_fork -x test_multiprocessing_forkserver -x test_multiprocessing_spawn
+        - ./venv/bin/python -m coverage run --pylib -m test --fail-env-changed -uall,-cpu -x test_multiprocessing_fork -x test_multiprocessing_forkserver -x test_multiprocessing_spawn -x test_concurrent_futures
```
I forgot the rationale for this?

tomMoral replied:
When coverage is used with multiprocessing, the spawning of new interpreter launch a new test session, resulting in a mess. I think it was previously working with the `fork` context but when I launch the test with the three backends, it also launches new test sessions.

Thus, I disabled coverage tests with test_concurrent_futures. I am not sure of what I should do if this is not the case. The duplicated test sessions could results from some command line arguments parsing either in the semaphore tracker or the forkserver but I do not think it is linked to this PR.

Let me know if this make sense.

pitrou replied:
It does make sense, thank you.

pitrou commented:
Thank you! This looks fine on the principle (apart from a couple small things mentioned in the review). You'll need to update the docs in `Doc/library/concurrent.futures.rst`.

tomMoral committed:
CLN contex->mp_context+FIX context in test

- Rename context to mp_context in ProcessPoolExecutor constructor
- Fix the context used in test_interpreter_shutdown

tomMoral committed:
TST add test_ressources_gced_in_workers

- Ensure that the job argument passed are freed asap

tomMoral committed:
DOC update ProcessPoolExecutor doc

pitrou commented on Doc/library/concurrent.futures.rst:
```diff
@@ -191,20 +191,29 @@ that :class:`ProcessPoolExecutor` will not work in the interactive interpreter.
    given, it will default to the number of processors on the machine.
    If *max_workers* is lower or equal to ``0``, then a :exc:`ValueError`
    will be raised.
+   *mp_context* can be a multiprocessing context or any object providing a
```
I'd rather restrict it to a multiprocessing context. Perhaps we'll use other multiprocessing APIs in the future.

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -55,6 +56,15 @@ def my_method(self):
 
+class EventfulGCObj():
+    def __init__(self, ctx):
+        mgr = get_context(ctx).Manager()
```
What is the rationale for using a manager here? Since the object is instantiated in the parent, it should be inheritable by the child anyway.

tomMoral replied:
The executor is launched in the `setup` method of the `TestCase`. Thus, there is no possibility to pass the `Event` object thru inheritence and the job `(id, obj)` is passed via pickle, which requires the `Manager`.

pitrou replied:
Oh, right. I've never used managers and I was surprised to see this...

pitrou commented on Lib/test/test_concurrent_futures.py:
```diff
@@ -591,6 +674,34 @@ def test_traceback(self):
+        future = self.executor.submit(id, obj)
+        future.result()
+
+        assert obj.event.wait(timeout=1)
```
By convention, we'd use `self.assertTrue` (also so that assertions are still checked if running with `-O`).

tomMoral committed:
FIX doc context + use assertTrue

pitrou committed:
Doc nit: simplify sentence

pitrou committed:
NEWS nits

pitrou commented:
Thank you @tomMoral !

tomMoral commented:
Thank @pitrou 
I will follow up with the other PR from #1013 asap.
